### PR TITLE
One hot encoding

### PR DESCRIPTION
**Pull request by iamzoltan**
Just added one hot encoding representation of the board. I broke the diagonal part of the heuristic, and removed the saved_state functionality. Please comment and help me improve where I can!

### CODE REVIEW

**Comment by goodboy** [comment]
@iamzoltan looks mostly good to me.

It seems this PR is a docs pump yah?
Not sure if you're looking for a whole rundown or not.
The docs improvements definitely be drippin :surfer:

File: connect_four/main.py [hunk 1]
```diff
@@ -1,30 +1,35 @@
...
+def user_move(board: np.ndarray, _player: BoardPiece) -> np.int8:
+    """
     Take player input and return action if it is a legal one.
-    '''
+
+    Parameters:
+        board: an array representation of the board
+        _player: user player
+
+    Returns:
```
Comment (goodboy) [inline connect_four/main.py:19]: I will say "naming" the return value comes off a bit weird where you don't need the assignment an can just return the output of another call.

Not sure it's that big a deal though if it adds clarity.

File: connect_four/agents/common.py [hunk 2]
```diff
@@ -221,10 +267,18 @@ def check_end_state(
...
+        player: current player we are using to check the game state
+        last_action: last action of player, used for faster searches
+
+    Returns:
+        GameState: current state of the game. (Win, Draw, Ongoing)
     """
     if connected_four(board, player):
         return GameState.IS_WIN
-    elif np.count_nonzero(board) == 42:
+    elif np.count_nonzero(board[:, :, 0]) == 0:
```
Comment (goodboy) [inline connect_four/agents/common.py:281]: Is this a flatten operation?

File: connect_four/main.py [hunk 3]
```diff
@@ -37,16 +42,28 @@ def human_vs_agent(
...
+        generate_move_2: arguments for the generate_move function
+        player_1: name of player 1
+        player_2: name of player 2
+        args_1: additional arguments for the GenMove function
+        args_2: additional arguments for the GenMove function
+        init_1: placeholder function used in prepcompiling to speed up calculations
+        init_2: placeholder function used in prepcompiling to speed up calculations
+
+    Returns:
+        None
```
Comment (goodboy) [inline connect_four/main.py:60]: I feel like some of this gets a bit duplicate with the type annotations.

File: connect_four/agents/common.py (hunk 2 above)
Comment (iamzoltan) [reply connect_four/agents/common.py:281]: what do you mean?

File: connect_four/agents/common.py [hunk 4]
```diff
@@ -173,38 +212,45 @@ def connected_four(
...
                 return True
-            block = board[i:i+num_to_win, j:j+num_to_win]
             # check \ diagonal
-            diag = np.diag(block)
+            diag = board[[i, i+1, i+2, i+3], [j, j+1, j+2, j+3]]
             if (diag == player).all():
                 return True
             # check / diagonal
-            diag_flip = np.diag(block[::-1, :])
+            diag_flip = board[[i+3, i+2, i+1, i], [j, j+1, j+2, j+3]]
```
[the remark this answers is not in the document]
Comment (iamzoltan) [reply connect_four/agents/common.py:241]: yea I think you are right, ill change that

**Comment by iamzoltan** [comment]
@goodboy Is also a change from using a 2d array with np.int8 to represent a board piece to a 3d boolean array using one hot encoding to represent board pieces

File: connect_four/agents/common.py (hunk 2 above)
Comment (goodboy) [reply connect_four/agents/common.py:281]: Disregard, saw your note about 3d board.
I was thinking 2d and was confused by the indexing - wasn't sure if that was some syntax for flattening to 1d that I didn't know :exploding_head:.

Makes sense now.

**Comment by goodboy** [comment]
>from using a 2d array with np.int8 to represent a board piece to a 3d boolean array

@iamzoltan stick that explanation in your PR description next time too :wink:

**Comment by goodboy** [comment]
Ahh so this is like a "2d board tensor" now yah?

File: connect_four/agents/common.py (hunk 2 above)
Comment (iamzoltan) [reply connect_four/agents/common.py:281]: sall good, I feel like this representation give it a pretty big speed boost. the minimax tree can search at a depth of 5 pretty fast now

File: connect_four/main.py (hunk 1 above)
Comment (iamzoltan) [reply connect_four/main.py:19]: will remove it

File: connect_four/main.py (hunk 3 above)
Comment (iamzoltan) [reply connect_four/main.py:60]: agreed, ill get rid of the duplicate comments

**Comment by iamzoltan** [comment]
@goodboy yup, like a 2d tensor board.

@ikhwankhalid just pushed the latest changes

will merge into master so we can get the MCTS working with the new baord asap. let me know what I should tweak before hand.

**Comment by ikhwankhalid** [comment]
Looks good now

**Pull request merged**